### PR TITLE
Ensure CI fails if tests fail

### PR DESCRIPTION
**Pull request by tomhjp**
Adding all of GitHub's suggested reviewers, but feel free to forward me on to other reviewers.

I noticed in #84 that the CI passes even if there are failures in the build and/or test, see here for an example on main: https://github.com/hashicorp/go-secure-stdlib/actions/runs/5602106840. See here for an example of the changes from this PR: https://github.com/hashicorp/go-secure-stdlib/actions/runs/5962737784

While we could probably fix it up with some bash wizardry, I also thought it seemed worthwhile to allow submodules to have their own Go version requirements, so I made a matrix strategy and also added an introspecting `verify-ci` step that ensures people don't forget to add their own module when they create new modules.

Finally, there are a few minor fixes, mostly where a module was actually depending on some Go features from a version ahead of its go.mod `go` directive. I've run `go mod tidy` in each module that required a `go` directive upgrade.

The build tags I added in `listenerutil` are not required from Go 1.18, but I thought it was less disruptive to add them than to bump the required Go version up further.

### CODE REVIEW

**Review by johanbrandhorst — APPROVED** [review]
LGTM, thanks!

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -1,24 +1,46 @@
...
-      run: find . -name go.mod -execdir go test ./... \;
\ No newline at end of file
+      run: cd ${{ matrix.module }} && go test ./...
+
+  # Verify that every module folder is mentioned in this workflow file.
+  verify-ci:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+    - run: find * -name go.mod -print0 | xargs -0 dirname | xargs -I % grep % .github/workflows/go.yml > /dev/null
```
[the remark this answers is not in the document]
Comment (tomhjp) [reply .github/workflows/go.yml]: Good call! Added in ed47383aba798851b667af8a9e5a610849f12f73

**Review by johanbrandhorst — APPROVED** [review]

**Comment by tomhjp** [comment]
🤦 sorry for dismissing the approval again - it was such a tiny change too 😅

**Review by jefferai — APPROVED** [review]

**Comment by tomhjp** [comment]
Thanks!

**Pull request merged**